Unexpected test runs count with retry analyzer

### TestNG Version
7.6.1

### Expected behavior
tests run count must depend only on parameters size and rules defined by retry analyzer 


### Actual behavior
tests run count depends on parallel flag of dataProvider annotation 

### Is the issue reproducible on runner?

- [ ] Shell
- [x] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [x] IntelliJ
- [ ] NetBeans


### Test case sample
```

import org.testng.IRetryAnalyzer;
import org.testng.ITestResult;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Arrays;
import java.util.Iterator;

public class RATest {


    public static class RA implements IRetryAnalyzer {

        int cnt = 0;
        static final int threshold = 2;


        @Override
        public synchronized boolean retry(ITestResult result) {
            System.out.println("retry:[" + hashCode() + "](" + Arrays.toString(result.getParameters()) + ")" + Thread.currentThread().getId());
            return cnt++ < threshold;
        }
    }


    @DataProvider(name = "test data", parallel = true) // parallel = false
    public Iterator<String> data() {
        return Arrays.asList("1", "2").iterator();
    }


    @Test(dataProvider = "test data", retryAnalyzer = RA.class)
    public void foo(String ignored) throws IOException {
        throw new FileNotFoundException("this can be retried");
    }


}

```
here is the output for ```parallel=true``` (5 times)
```
retry:[356127017]([1])16
retry:[1514394209]([2])17 - a new instance of retry analyzer !
retry:[356127017]([2])17
retry:[356127017]([1])16
retry:[356127017]([2])17

```
[ERROR] Tests run: 5, Failures: 2, Errors: 0, Skipped: 3, Time elapsed: 3.891 s <<< FAILURE! - in TestSuite
[ERROR] Tests run: 5, Failures: 2, Errors: 0, Skipped: 3

here is the output for ```parallel=false``` (6 times)
```
retry:[1615523464]([1])15
retry:[1615523464]([1])15
retry:[1615523464]([1])15
retry:[132034104]([2])15
retry:[132034104]([2])15
retry:[132034104]([2])15

```
[ERROR] Tests run: 6, Failures: 2, Errors: 0, Skipped: 4, Time elapsed: 3.379 s <<< FAILURE! - in TestSuite
[ERROR] Tests run: 6, Failures: 2, Errors: 0, Skipped: 4

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -5,6 +5,7 @@ New: Ability to provide custom error message for assertThrows\expectThrows metho
 Fixed: GITHUB-2780: Use SpotBugs instead of abandoned FindBugs
 Fixed: GITHUB-2801: JUnitReportReporter is too slow
 Fixed: GITHUB-2807: buildStackTrace should be fail-safe
+Fixed: GITHUB-2798: Parallel executions coupled with retry analyzer results in duplicate retry analyzer instances being created (Krishnan Mahadevan)
 
 7.6.1
 Fixed: GITHUB-2761: Exception: ERROR java.nio.file.NoSuchFileException: /tmp/testngXmlPathInJar-15086412835569336174 (Krishnan Mahadevan)

--- a/testng-core/src/main/java/org/testng/internal/BaseTestMethod.java
+++ b/testng-core/src/main/java/org/testng/internal/BaseTestMethod.java
@@ -58,7 +58,7 @@ public abstract class BaseTestMethod implements ITestNGMethod, IInvocationStatus
   private boolean m_enabled;
 
   private final String m_methodName;
-  // If a depended group is not found
+  // If a depends on group is not found
   private String m_missingGroup;
   private String m_description = null;
   protected AtomicInteger m_currentInvocationCount = new AtomicInteger(0);
@@ -449,10 +449,6 @@ public abstract class BaseTestMethod implements ITestNGMethod, IInvocationStatus
     return m_annotationFinder;
   }
 
-  protected IClass getIClass() {
-    return m_testClass;
-  }
-
   static StringBuilder stringify(String cls, ConstructorOrMethod method) {
     StringBuilder result = new StringBuilder(cls).append(".").append(method.getName()).append("(");
     return result.append(method.stringifyParameterTypes()).append(")");
@@ -788,16 +784,13 @@ public abstract class BaseTestMethod implements ITestNGMethod, IInvocationStatus
       return this.m_retryAnalyzer;
     }
     final String keyAsString = getSimpleName() + "#" + getParameterInvocationCount();
-    final IRetryAnalyzer currentRetryAnalyzerInMap = m_testMethodToRetryAnalyzer.get(keyAsString);
-    if (currentRetryAnalyzerInMap != null) {
-      return currentRetryAnalyzerInMap;
-    }
-    BasicAttributes ba = new BasicAttributes(null, this.m_retryAnalyzerClass);
-    CreationAttributes attributes = new CreationAttributes(tr.getTestContext(), ba, null);
-    IRetryAnalyzer instance =
-        (IRetryAnalyzer) Dispenser.newInstance(m_objectFactory).dispense(attributes);
-    m_testMethodToRetryAnalyzer.put(keyAsString, instance);
-    return instance;
+    return m_testMethodToRetryAnalyzer.computeIfAbsent(
+        keyAsString,
+        key -> {
+          BasicAttributes ba = new BasicAttributes(null, this.m_retryAnalyzerClass);
+          CreationAttributes attributes = new CreationAttributes(tr.getTestContext(), ba, null);
+          return (IRetryAnalyzer) Dispenser.newInstance(m_objectFactory).dispense(attributes);
+        });
   }
 
   private static boolean isNotParameterisedTest(ITestResult tr) {

--- a/testng-core/src/test/java/test/retryAnalyzer/RetryAnalyzerTest.java
+++ b/testng-core/src/test/java/test/retryAnalyzer/RetryAnalyzerTest.java
@@ -8,6 +8,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.function.Function;
 import java.util.stream.Collectors;
 import org.testng.Assert;
 import org.testng.ITestResult;
@@ -22,7 +23,6 @@ import test.InvokedMethodNameListener;
 import test.SimpleBaseTest;
 import test.retryAnalyzer.dataprovider.issue2163.TestClassPoweredByDataProviderSample;
 import test.retryAnalyzer.github1519.MyListener;
-import test.retryAnalyzer.github1519.TestClassSample;
 import test.retryAnalyzer.github1600.Github1600Listener;
 import test.retryAnalyzer.github1600.Github1600TestSample;
 import test.retryAnalyzer.github1706.DataDrivenSample;
@@ -38,8 +38,20 @@ import test.retryAnalyzer.issue1946.RetryAnalyzer;
 import test.retryAnalyzer.issue1946.TestclassSample1;
 import test.retryAnalyzer.issue1946.TestclassSample2;
 import test.retryAnalyzer.issue2684.SampleTestClassWithGroupConfigs;
+import test.retryAnalyzer.issue2798.HashCodeAwareRetryAnalyzer;
+import test.retryAnalyzer.issue2798.TestClassSample;
 
 public class RetryAnalyzerTest extends SimpleBaseTest {
+
+  @Test(description = "GITHUB-2798")
+  public void ensureNoDuplicateRetryAnalyzerInstancesAreCreated() {
+    create(TestClassSample.class).run();
+    Map<Integer, Long> collected =
+        HashCodeAwareRetryAnalyzer.hashCodes.stream()
+            .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+    assertThat(collected.keySet()).hasSize(1);
+  }
+
   @Test
   public void testInvocationCounts() {
     TestNG tng = create(InvocationCountTest.class);
@@ -61,10 +73,10 @@ public class RetryAnalyzerTest extends SimpleBaseTest {
 
   @Test
   public void testIfRetryIsInvokedBeforeListener() {
-    TestNG tng = create(TestClassSample.class);
+    TestNG tng = create(test.retryAnalyzer.github1519.TestClassSample.class);
     tng.addListener(new MyListener());
     tng.run();
-    assertThat(TestClassSample.messages)
+    assertThat(test.retryAnalyzer.github1519.TestClassSample.messages)
         .containsExactly("afterInvocation", "retry", "afterInvocation");
   }
 

--- a/testng-core/src/test/java/test/retryAnalyzer/issue2798/HashCodeAwareRetryAnalyzer.java
+++ b/testng-core/src/test/java/test/retryAnalyzer/issue2798/HashCodeAwareRetryAnalyzer.java
@@ -1,0 +1,20 @@
+package test.retryAnalyzer.issue2798;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.testng.IRetryAnalyzer;
+import org.testng.ITestResult;
+
+public class HashCodeAwareRetryAnalyzer implements IRetryAnalyzer {
+
+  public static final List<Integer> hashCodes = new ArrayList<>();
+
+  int cnt = 0;
+  static final int threshold = 2;
+
+  @Override
+  public synchronized boolean retry(ITestResult result) {
+    hashCodes.add(this.hashCode());
+    return cnt++ < threshold;
+  }
+}

--- a/testng-core/src/test/java/test/retryAnalyzer/issue2798/TestClassSample.java
+++ b/testng-core/src/test/java/test/retryAnalyzer/issue2798/TestClassSample.java
@@ -1,0 +1,21 @@
+package test.retryAnalyzer.issue2798;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Iterator;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+public class TestClassSample {
+
+  @DataProvider(name = "test data", parallel = true)
+  public Iterator<Object[]> data() {
+    return Arrays.asList(new Object[] {"1"}, new Object[] {"2"}).iterator();
+  }
+
+  @Test(dataProvider = "test data", retryAnalyzer = HashCodeAwareRetryAnalyzer.class)
+  public void foo(String ignored) throws IOException {
+    throw new FileNotFoundException("this can be retried");
+  }
+}

